### ISSUE TITLE
Implementing "ngDoCheck" as async causes an infinite loop

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/angular#1375]
_(Internal bug is [b/109662261](b/109662261))_

An internal customer writes:

> Overriding `ngDoCheck` with an `async` implementation causes an infinite loop when compiled with dart2js but not with DDC, even if the method has no side effects.
> 
> For example:
> 
> ```dart
> @Component(selector: 'my-component')
> class MyComponent implements DoCheck {
>   @override
>   void ngDoCheck() async {
>     print('checking');
>   }
> }
> ```
> 
> ...causes an infinite stream of 'checking' messages and locks up the tab at the point the component should be rendered but before any content is rendered.

Things to verify:

* [x] Why does this only occur in Dart2JS (is this `DART_CODEGEN_MODE` related?)
* [x] How can we mitigate this at compile-time
* [x] How can we mitigate this at runtime

... we might want to update the change detection FAQ/guide as well @hterkelsen

**Comment by matanlurey** [comment]
The following test does _not_ fail by default in Chrome+DDC+Debug mode:

```dart
@TestOn('browser')
import 'package:angular/angular.dart';
import 'package:angular_test/angular_test.dart';
import 'package:test/test.dart';

import '1375_infinite_docheck_test.template.dart' as ng;

void main() {
  test('should not cause an infinite loop', () async {
    final testBed = NgTestBed.forComponent<TestAsyncDoCheck>(
      ng.TestAsyncDoCheckNgFactory,
    );
    await testBed.create();
  });
}

@Component(
  selector: 'test-async-do-check',
  template: '<has-async-do-check></has-async-do-check>'
)
class TestAsyncDoCheck {}

@Component(
  selector: 'has-async-do-check',
  template: '',
)
class HasAsyncDoCheck implements DoCheck {
  @override
  void ngDoCheck() async {}
}
```

... will try next with debug mode disabled, and then in Dart2JS.

**Comment by matanlurey** [comment]
With debug mode disabled, it still does nothing bad. I'll investigate further.

**Comment by matanlurey** [comment]
I can't get that test case to fail in either DartDevCompiler or Dart2JS yet.

**Comment by matanlurey** [comment]
Here is the code compiled with DDC, approximately:

```js
// 1375_infinite_docheck_test.ddc.js
class HasAsyncDoCheck extends core.Object {
  ngDoCheck() {
    return async.async(dart.dynamic, function* ngDoCheck() {});
  }
}
```

It was tough to find the definition of `async.async`, but I traced it to:

```js
async.async = function(T, initGenerator) {
    let iter = null;
    let onValue = null;
    let onError = null;
    function onAwait(value) {
      let f = null;
      if (async._Future.is(value)) {
        f = value;
      } else if (async.Future.is(value)) {
        f = new async._Future.new();
        async._Future._chainForeignFuture(value, f);
      } else {
        f = new async._Future.value(value);
      }
      f = f[_thenNoZoneRegistration](core.Object, onValue, onError);
      return f;
    }
    onValue = value => {
      let iteratorResult = iter.next(value);
      value = iteratorResult.value;
      return iteratorResult.done ? value : onAwait(value);
    };
    onError = value => {
      let iteratorResult = iter.throw(value);
      value = iteratorResult.value;
      return iteratorResult.done ? value : onAwait(value);
    };
    let zone = async.Zone.current;
    if (!dart.equals(zone, async.Zone.root)) {
      onValue = zone.registerUnaryCallback(core.Object, core.Object, onValue);
      onError = zone.registerUnaryCallback(core.Object, core.Object, onError);
    }
    let asyncFuture = new (async._Future$(T)).new();
    let body = () => {
      try {
        iter = initGenerator()[Symbol.iterator]();
        let iteratorValue = iter.next(null);
        let value = iteratorValue.value;
        if (iteratorValue.done) {
          if (async.Future.is(value)) {
            if (async._Future.is(value)) {
              async._Future._chainCoreFuture(value, asyncFuture);
            } else {
              async._Future._chainForeignFuture(value, asyncFuture);
            }
          } else {
            asyncFuture[_completeWithValue](value);
          }
        } else {
          async._Future._chainCoreFuture(onAwait(value), asyncFuture);
        }
      } catch (e) {
        let s = dart.stackTrace(e);
        if (dart.test(dart.startAsyncSynchronously)) {
          async.scheduleMicrotask(() => {
            async._completeWithErrorCallback(asyncFuture, e, s);
          });
        } else {
          async._completeWithErrorCallback(asyncFuture, e, s);
        }
      }
    };
    dart.fn(body, VoidToNull());
    if (dart.test(dart.startAsyncSynchronously)) {
      body();
    } else {
      async.scheduleMicrotask(body);
    }
    return asyncFuture;
  };
```

Compare with Dart2JS:

```js
// 1375_infinite_docheck_test.dart.js
  Z.HasAsyncDoCheck.prototype = {
    ngDoCheck$0: function() {
      var $async$goto = 0, $async$completer = P._makeAsyncAwaitCompleter(null);
      var $async$ngDoCheck$0 = P._wrapJsFunctionForAsync(function($async$errorCode, $async$result) {
        if ($async$errorCode === 1)
          return P._asyncRethrow($async$result, $async$completer);
        while (true)
          switch ($async$goto) {
            case 0:
              // Function start
              // implicit return
              return P._asyncReturn(null, $async$completer);
          }
      });
      return P._asyncStartSync($async$ngDoCheck$0, $async$completer);
    }
  };
```

**Comment by matanlurey** [comment]
Here is a minimal program compiled with DDC versus Dart2JS:

```dart
import 'dart:async';

import 'package:meta/dart2js.dart' as dart2js;

void main() {
  runZoned(() {
    new AsyncDoEvent().doEvent();
  }, zoneSpecification: new ZoneSpecification(
    scheduleMicrotask: (self, parent, zone, fn) {
      print('scheduleMicrotask()');
      return parent.scheduleMicrotask(zone, fn);
    },
  ));
}

abstract class HasEvent {
  void doEvent();
}

class AsyncDoEvent implements HasEvent {
  @dart2js.noInline
  void doEvent() async {
    print('doEvent()');
  }
}
```

## DDC

```
doEvent()
```

### Dart2JS

```
doEvent()
scheduleMicrotask()
```

... so the main difference here is Dart2JS schedules another zone turn (which, due to the semantics of `DoCheck`, starves the micro-task queue), and DDC does not. Note that any non-trivial use (i.e. using `await` or Futures) in `doCheck()` should also reproduce this bug in DDC, not just Dart2JS.

**Comment by leonsenft** [comment]
Would running `ngDoCheck()` _outside_ of the Angular zone prevent the cycle caused by scheduling microtasks?

**Comment by matanlurey** [comment]
The customer was not seeing infinite loops on less-trivial `async` functions. So I tried:

```dart
import 'dart:async';

import 'package:meta/dart2js.dart' as dart2js;

void main() {
  runZoned(() {
    new AsyncDoEvent().doEvent();
  }, zoneSpecification: new ZoneSpecification(
    scheduleMicrotask: (self, parent, zone, fn) {
      print('scheduleMicrotask()');
      return parent.scheduleMicrotask(zone, fn);
    },
  ));
}

abstract class HasEvent {
  void doEvent();
}

class AsyncDoEvent implements HasEvent {
  @dart2js.noInline
  void doEvent() async {
    print('doEvent()');
    if (alwaysFalse()) {
      await null;
    }
  }
}

@dart2js.noInline
bool alwaysFalse() => 'a'.contains('b');
```

## DDC

```
doEvent()
```

## Dart2JS

```
doEvent()
scheduleMicrotask()
```

**Comment by matanlurey** [comment]
@leonsenft:

> Would running `ngDoCheck()` _outside_ of the Angular zone prevent the cycle...

Yes, though it could also cause other user behavior to break, unfortunately. I'm still trying to determine why in some cases this works fine (the user does not get an infinite time out) and in a trivial case it does.

**Comment by matanlurey** [comment]
If the parent is OnPush, there is no infinite loop, I've observed.

**Comment by matanlurey** [comment]
OK, so I think I've narrowed down that it does crash, no matter what, in Dart2JS.

We have a few options here (and they aren't all mutually exclusive):

## Issue a compile error for `ngDoCheck()` if `async`

This _won't_ catch other usages of micro-tasks/timers (explicitly or as a side-effect), but I'm guessing it will catch _some_ (since `async` is so used), and at least the error message could strongly warn against any asynchronous code in `DoCheck`.

## Throw an error if an asynchronous side-effect is present in `ngDoCheck`

We could do this only in DDC, if we wanted. It would require code like the following:

```dart
void runDoCheckSafe(DoCheck directive) {
  if (isDevMode) {
    return _assertNoAsync(directive.ngDoCheck);
  }
  directive.ngDoCheck();
}

void _assertNoAsync(void Function() fn) {
  // Run in a zone that throws if microtasks or timers are scheduled?
}
```

... unfortunately this would not have caught the above examples, because DDC isn't scheduling any microtasks for an `async` function that does not use `await` (unless Dart2JS). I'm not sure that is intentional or not, yet.

## Run `ngDoCheck` outside of AngularDart, at all times

Basically, @leonsenft's suggestion above:

```dart
void runDoCheckSafe(DoCheck directive) {
  ChangeDetectionHost.runOutside(directive.ngDoCheck);
}
```

... we'd have to add `runOutside` as an internal method on `ChangeDetectionHost`. At minimum, this would avoid the problem. It could cause other problems (users would expect changes that never happen), but at least it doesn't crash the browser and would be consistent across web compilers.

**Comment by matanlurey** [comment]
@natebosch:

> The custom zone that disallows scheduling microtasks seems ok to me. Why would we only be able to do it with DDC?

A few reasons:

1. `DoCheck` is expected to be fast, i.e. it is used for `ngIf` and `ngFor`. The overhead of swapping zones for every call to `ngDoCheck()` may (visibly) negatively effect performance. At best we'd need to make it a flag and have some large client apps verify in their QA environment.

2. Even if we did this, it wouldn't catch the cases above, because DDC never schedules one.

**Comment by matanlurey** [comment]
If we had the infrastructure for developing a custom lint, i.e. `@strictlySynchronous`, then I'd probably try to use that (i.e. don't allow `async` or to invoke code that returns a `Future`, `Stream`, `Timer`, etc). It definitely would be difficult to write such a lint, in any case.

**Comment by har79** [comment]
A custom zone that disallows scheduling microtasks would prevent some code that does work so, if that is implemented, could an opt-out be added (`@allowFutures` or similar)? I fixed our use of this by restricting futures to only be used when the state has actually changed, for example:

```
void ngDoCheck() {
  if (identical(_old, _new)) return;
  _old = _new;
  _refresh();
}

Future<void> _refresh() { ... }
```

This will still cause another change detection when the async code is called but that cycle will return immediately.

I recognise this is still potentially dangerous as future changes could accidentally reintroduce the error. But I don't want it to be disallowed entirely because there isn't always another way to implement the same functionality, e.g., one of our components is doing similar work to ngIf/ngFor but with some delayed execution.

**Comment by matanlurey** [comment]
@har79 Have you considered using `ngAfterChanges` for this, instead of `DoCheck`?

That would be the appropriate use. `DoCheck` is supposed to only be used for change _detection_.

**Comment by har79** [comment]
I initially looked at AfterChanges but both of the components/directives
*are* doing change detection (my example may have been too stripped down).

One of them has a mutable object as a binding and needs to update if the
content of the object changes (so equality-based, not identity-based), and
wouldn't be picked up by the default change detection.

The other directive should only update if a specific binding changes and
should ignore the others. So it *could* allow the default change detection
but would then have to redo its own afterwards.

**Comment by matanlurey** [comment]
Yeah I'm not sure we have a story for you yet.

You could obviously go with the guarded case (without using `async`). It will be dangerous though.

**Comment by matanlurey** [comment]
At this point, we are going to just lint `async` methods here. @har79 sent you an internal CL.

**Comment by matanlurey** [comment]
Closed at HEAD.
